### PR TITLE
Series File Tombstone Tracking

### LINKED ISSUE #9514: Tombstone aggregate tracking
Add ability to track tombstone aggregates on series files segments to determine if we need to compact them.

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds series & tombstone statistic tracking to the series file compactor. These are calculated on a per-segment basis to later determine if they are past a threshold for compacting.

Fixes #9514.

### CODE REVIEW

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/210506072/src/github.com/influxdata/influxdb/query/monitor_test.go:49:11: this value of err is never used (SA4006)
/tmp/210506072/src/github.com/influxdata/influxdb/tsdb/index/tsi1/log_file.go:139:2: 'if err != nil { return err }; return nil' can be simplified to 'return err' (S1013)
```

File: tsdb/series_partition.go [hunk 1]
```diff
@@ -614,6 +648,9 @@ func (c *SeriesPartitionCompactor) compactIndexTo(index *SeriesIndex, seriesN ui
 		return err
 	}
 
+	// Attach stats to compactor.
+	c.Stats = stats
+
 	return nil
 }
```
Comment (benbjohnson) [inline tsdb/series_partition.go]: Stats are built using the current index snapshot to calculate tombstones. Orphaned tombstones are calculated based on a bitset to determine if a tombstone's series has already been compacted away in an segment.

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
``````

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/795134988/src/github.com/influxdata/influxdb/query/monitor_test.go:49:11: this value of err is never used (SA4006)
```

**Comment by e-dard** [comment]
@benbjohnson `hercules-influx` with an empty comment is a bug in the static checker. Sometimes it means that the branch doesn't build.

**Comment by e-dard** [comment]
@benbjohnson I don't know why it's not emitting the error. I will have to look at it. The issue anyway is:

```
⇒  megacheck github.com/influxdata/influxdb/...
query/monitor_test.go:49:11: this value of err is never used (SA4006)
```

File: tsdb/series_partition.go [hunk 2]
```diff
@@ -389,6 +393,13 @@ func (p *SeriesPartition) compactionsEnabled() bool {
 	return p.compactionsDisabled == 0
 }
 
+// SeriesPartitionStats returns stats generated by compaction.
+func (p *SeriesPartition) Stats() *SeriesPartitionStats {
+	p.mu.Lock()
```
Comment (e-dard) [inline tsdb/series_partition.go]: `RLock` / `RUnlock`?

Comment (jwilder) [reply tsdb/series_partition.go]: +1

File: tsdb/series_partition.go [hunk 3]
```diff
@@ -56,6 +58,7 @@ func NewSeriesPartition(id int, path string) *SeriesPartition {
 		CompactThreshold: DefaultSeriesPartitionCompactThreshold,
 		Logger:           zap.NewNop(),
 		seq:              uint64(id) + 1,
+		stats:            NewSeriesPartitionStats(),
```
Comment (jwilder) [inline tsdb/series_partition.go]: What happens on restart/re-open?  Are any stats accumulated prior to the restart lost?

**Comment by e-dard** [comment]
@benbjohnson what is the status of this PR? Is there an advantage to getting it into the next release? /cc @rbetts

**Comment by benbjohnson** [comment]
@e-dard I need to update based on the feedback but it's been low priority compared to EARs. There's no benefit to adding this to the next release.

**Comment by e-dard** [comment]
don't close.

**Comment by danxmoran** [comment]
@benbjohnson working on clearing our old PR queue, is this something that would make sense to re-target at 2.x?

**Pull request closed without merging**